README.rst

```diff
@@ -1,102 +1,208 @@
+.. _easy-to-override hygiene: examples/anaphoric-if.kl
+.. _Custom languages: examples/rpn.kl
+.. _Syntax objects: TODO: link to a short example which explains that in
+   Racket, syntax objects are introduced via ``#'(...)``, whereas in Klister
+   they are introduced via ``'(...)``. Also explain that Klister does not have
+   unannotated s-expressions.
+.. _module system: TODO: write a short example demonstrating what can be
+   expressed with the import and export primitives.
+
+Features we borrow from Haskell:
+
+* Statically-typed; we check that macro implementations have type
+  ``(-> Syntax (Macro Syntax)``, and *then* that the generated code is also
+  well-typed. We thus aim for the expressivity of Template Haskell, not the
+  extra guarantees of Typed Template Haskell.
+* Purely functional; primitives with side-effects either run in the ``IO``
+  monad or in the ``Macro`` monad, as appropriate.
+* Higher-kinded types; for example `monads`_ are defined as a library.
+* `Algebraic datatypes`_. In Racket, ``(list 1 2 3)`` and ``'(1 2 3)``
+  construct the same value, an s-expression which satisfies the ``list?``
+  predicate. In Klister, the former produces a value of the algebraic datatype
+  ``(List Integer)``, while the latter produces a value of type ``Syntax``.
+  This syntax can be "opened", via ``(open-syntax '(1 2 3))``, to obtain the
+  value ``(list-contents (list '1 '2 '3))`` of type
+  ``(Syntax-Contents Syntax)``.
+* `Bidirectional type inference`_; the type of an argument can specialize the
+  type of the function being called.
+* Bidirectional type inference; the type of a function can specialize the type
+  of its arguments.
+
+.. _monads: TODO: link to monad.kl's Monad definition, and add a comment there
+   highlighting the inferred type, especially the higher-kinded type variable.
+.. _Algebraic datatypes: TODO: write a small example defining and matching on
+   an algebraic type. Perhaps Either?
+.. _Bidirectional type inference: TODO: write a small example demonstrating
+   those two features.
+
+Features which make Klister special (but not *unique*; see the `bibliography`_):
+
+* `Type-providing macros`_; a macro used in an argument position can provide
+  type information about the argument, and can thus specialize the type of the
+  function being called.
+* `Type-aware macros`_; a macro used in an argument position can obtain the
+  type of the argument it needs to generate, and thus the type of a function
+  can affect what arguments are generated.
+* `Stuck macros`_; it is thus possible for multiple macros to affect what each
+  other generates. The language primitives are designed so that the order in
+  which the macros are expanded cannot affect their results, and indeed the
+  same is true for the order in which the macro expansion and type-inference
+  steps are interleaved. This makes Klister code more robust to refactoring.
+* `Problem-aware macros`_; in addition to the type, a macro can learn which
```

Comment:
I'd say that each problem corresponds to a form of judgment in the specification of the language. Because each form of judgment regulates the validity of a particular syntactic construct (be it type, expression, declaration, etc), these two things correspond.

Reply:
I like the idea, but we didn't write down those judgments, so there currently isn't anything to correspond to. How about "would correspond to a form of judgment if the language was formalized, e.g. a typing judgment for the `expression` problem and a well-formed type judgment for the `type` problem"?

Reply:
I attempted this counter-proposal in fb2bd753dbe0cb2198a4236f9f182802b1464c9d, let me know what you think!